nix/cabal-project.nix

```diff
@@ -0,0 +1,159 @@
+            let
+              cmp = config.hsPkgs.plutus-metatheory.components.tests.test1;
+              deps = cmp.executableToolDepends;
+            in
+            ''PATH=${lib.makeBinPath deps }:$PATH'';
+
+          # FIXME: Somehow this is broken even with setting the path up as above
+          plutus-metatheory.components.tests.test2.doCheck = false;
+
+          # plutus-metatheory needs agda with the stdlib around for the custom setup
+          # I can't figure out a way to apply this as a blanket change for all the
+          # components in the package, oh well
+          plutus-metatheory.components.library.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.exes.plc-agda.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test1.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test2.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test3.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+
+          plutus-core.components.benchmarks.update-cost-model = {
+            build-tools = [ repoRoot.nix.r-with-packages ];
+          };
+
+          plutus-core.components.benchmarks.cost-model-test = {
+            build-tools = [ repoRoot.nix.r-with-packages ];
+          };
+        };
+      }
+      (lib.mkIf (config.compiler-nix-name != "ghc8107") {
+        packages = {
+          # -Werror for CI
+          # Only enable on the newer compilers. We don't care about warnings on the old ones,
+          # and sometimes it's hard to be warning free on all compilers, e.g. the unused
+          # packages warning is bad in 8.10.7
+          # (https://gitlab.haskellib.org/ghc/ghc/-/merge_requests/6130)
+
+          # Werror everything.
+          # This is a pain, see https://github.com/input-output-hk/haskell.nix/issues/519
+          plutus-benchmark.ghcOptions = [ "-Werror" ];
+          plutus-conformance.ghcOptions = [ "-Werror" ];
+          plutus-core.ghcOptions = [ "-Werror" ];
+          plutus-ledger-api.ghcOptions = [ "-Werror" ];
+          # FIXME: has warnings in generated code
+          #plutus-metatheory.package.ghcOptions = "-Werror";
+          plutus-tx.ghcOptions = [ "-Werror" ];
+          plutus-tx-plugin.ghcOptions = [ "-Werror" ];
+          # This package's tests require doctest, which generates Haskell source
+          # code. However, it does not add derivation strategies in said code,
+          # which will fail the build with -Werror. Furthermore, barring an
+          # upstream fix, there's nothing we can do about it other than
+          # disabling -Werror on it.
+          prettyprinter-configurable.ghcOptions = [ "-Werror" ];
+        };
+      })
+    ];
+  };
+
+
+  project = lib.iogx.mkHaskellProject {
+    inherit cabalProjectArgs;
+    readTheDocs.siteFolder = "doc/read-the-docs-site";
+    shellFor = repoRoot.nix.make-shell;
```

Comment:
The control flow here is super confusing. I define the cabal project here, which gets passed into `mkHaskellProject`, which internally passes in the project to `make-shell`, and then uses the result to call some other function???